homework/shared_ptr/my_shared_ptr.hpp

```diff
@@ -14,94 +15,87 @@ class my_shared_ptr
 public:
     explicit my_shared_ptr(T* ptr) : ptr_(ptr)
     {
-        if (counter_ptr != nullptr)
-        {
-            counter_ptr->count_++;
-        }
-        else
-        {
             counter_ptr = new Counter();
             counter_ptr->count_++;
-        }
+            counter_ptr->weak_count++;
     };
 
     my_shared_ptr() : counter_ptr(new Counter()) { ptr_ = nullptr; };
 
-    my_shared_ptr(const my_shared_ptr<T>& other_shared_ptr)
+    my_shared_ptr(const my_shared_ptr<T>& other_shared_ptr) 
+        : ptr_(other_shared_ptr.ptr_),
+          counter_ptr(other_shared_ptr.counter_ptr)
     {
-        this->ptr_ = other_shared_ptr.ptr_;
-        this->counter_ptr = other_shared_ptr.counter_ptr;
-
         counter_ptr->count_++;
+        counter_ptr->weak_count++;
     };
 
-    my_shared_ptr(my_shared_ptr<T>&& ptr_moved)
+    my_shared_ptr(my_shared_ptr<T>&& ptr_moved) :  ptr_(ptr_moved.get())
     {
-        ptr_ = ptr_moved.get();
-        if (counter_ptr != nullptr)
-        {
-            counter_ptr->count_ = ptr_moved.use_count();
-        }
-        else
-        {
-            counter_ptr = new Counter();
-            counter_ptr->count_ = ptr_moved.use_count();
-        }
-        ptr_moved.reset(nullptr);
+            counter_ptr = ptr_moved.counter_ptr;
+            ptr_moved.ptr_ = nullptr;
     }
 
-    T& operator=(const my_shared_ptr<T>& other_shared_ptr)
+    my_shared_ptr<T>& operator=(const my_shared_ptr<T>& other_shared_ptr) 
     {
         std::cout << "operator move assigment" << "\n";
         delete ptr_;
         delete counter_ptr;
 
-        this->ptr_ = other_shared_ptr.ptr_;
-        this->counter_ptr = other_shared_ptr.counter_ptr;
+        ptr_ = other_shared_ptr.ptr_;
+        counter_ptr = other_shared_ptr.counter_ptr;
         if (other_shared_ptr)
         {
-            this->counter_ptr->count_++;
+            counter_ptr->count_++;
+            counter_ptr->weak_count++;
         }
+        return *this;
     };
 
-    T& operator=(const my_shared_ptr<T>&& some_ptr)
+    my_shared_ptr<T>& operator=(my_shared_ptr<T>&& some_ptr)
     {
-        if (some_ptr != this)
+        if (&some_ptr != this)
         {
             delete ptr_;
             delete counter_ptr;
 
-            this->ptr_ = some_ptr.ptr_;
-            this->counter_ptr = some_ptr.counter_ptr;
-            some_ptr.reset(nullptr);
+            ptr_ = some_ptr.ptr_;
+            counter_ptr = some_ptr.counter_ptr;
+            some_ptr = nullptr;
         }
         return *this;
     }
 
-    void reset(T* new_ptr)
+    void reset(T* new_ptr=nullptr)
     {
         delete ptr_;
         counter_ptr->count_ = 0;
```

Comment:
```suggestion
        counter_ptr->count_ = 1;
```
You will have one pointer pointing to it